### PR TITLE
Replace simpleMatchToFullName

### PR DESCRIPTION
**Pull request by romseygeek**
MappingLookup has a method simpleMatchToFieldName that attempts
to return all field names that match a given pattern; if no patterns match,
then it returns a single-valued collection containing just the pattern that
was originally passed in. This is a fairly confusing semantic.

This PR replaces simpleMatchToFullName with two new methods:
* getMatchingFieldNames(), which returns a set of all mapped field names
  that match a pattern.  Calling `getFieldType()` with a name returned by
  this method is guaranteed to return a non-null MappedFieldType
* getMatchingFieldTypes, that returns a collection of all MappedFieldTypes
  in a mapping that match the passed-in pattern. 

This allows us to clean up several call-sites because we know that 
MappedFieldTypes returned from these calls will never be null.  It also
simplifies object field exists query construction.

### CODE REVIEW

**Comment by romseygeek** [comment]
The test failure is a search integration test that asks for an object field in its stored field list and expects an error back.  With this PR we just ignore the field instead.  I think this is reasonable, but I can put the check for object fields back if need be.

File: server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java [hunk 1]
```diff
@@ -141,69 +138,22 @@ public static Query newFilter(SearchExecutionContext context, String fieldPatter
...
         BooleanQuery.Builder boolFilterBuilder = new BooleanQuery.Builder();
-        for (String field : fields) {
-            boolFilterBuilder.add(newFieldExistsQuery(context, field), BooleanClause.Occur.SHOULD);
+        for (MappedFieldType field : fields) {
+            boolFilterBuilder.add(field.existsQuery(context), BooleanClause.Occur.SHOULD);
         }
         return new ConstantScoreQuery(boolFilterBuilder.build());
     }
 
-    private static Query newFieldExistsQuery(SearchExecutionContext context, String field) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java:155]: Nice simplification!

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java [hunk 2]
```diff
@@ -140,27 +142,50 @@ private MappedFieldType getDynamicField(String field) {
...
      */
-    Set<String> simpleMatchToFullName(String pattern) {
+    Set<String> getMatchingFieldNames(String pattern) {
+        if ("*".equals(pattern)) {
+            return fullNameToFieldType.keySet();
+        }
         if (Regex.isSimpleMatchPattern(pattern) == false) {
             // no wildcards
-            return Collections.singleton(pattern);
+            return get(pattern) == null ? Collections.emptySet() : Collections.singleton(pattern);
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:180]: I think this changes the semantics slightly -- if you pass in a flattened subfield `my_flattened.subfield`, we'll that same value, whereas before we didn't. But if you pass in `my_flattened.sub*`, nothing will be returned. I think we should maintain the old behavior, which is more consistent.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java [hunk 3]
```diff
@@ -140,27 +142,50 @@ private MappedFieldType getDynamicField(String field) {
...
-     * Returns all the mapped field types.
+     * Returns all the mapped field types that match a pattern
+     *
+     * Note that if a field is aliased and both its actual name and its alias
+     * match the pattern, the returned collection will contain the field type
+     * twice.
      */
-    Collection<MappedFieldType> get() {
-        return fullNameToFieldType.values();
+    Collection<MappedFieldType> getMatchingFieldTypes(String pattern) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:151]: I think it'd be nice to keep a convenience method like `getAllFieldTypes`. It doesn't really add complexity and makes the call sites easier to understand.

**Review by jtibshirani — COMMENTED** [review]
This makes sense to me. I left a couple comments.

It seems okay to return nothing when requesting an object in stored fields. Are there any other places where this could cause user-facing changes?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:180]: ++ good catch, I'll change it back

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 3 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:151]: Really we want to remove `get()` as it's only used by the parent-join module to keep track of a singleton field type, so this felt like a convenient halfway house.  I can put it back in if you prefer though.

**Comment by romseygeek** [comment]
> Are there any other places where this could cause user-facing changes?

I don't think so, this was the only place that was explicitly checking for the existence of an object mapper.

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:180]: Turns out we do need this behaviour for flattened queries to work with query_string queries, and the inconsistent behaviour, although a bit odd, is necessary - and I think is what was happening before as well?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 3 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:151]: I don't feel strongly, but to me it helps. I also see it used in `MapperService`, and we might be able to replace some instances of `getMatchingFieldNames("*")` in tests?

File: server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java (hunk 2 above)
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java:180]: I managed to confuse myself earlier, I agree the current behavior is correct. The 'inconsistent behavior' I noted is actually just how flattened field resolution works, including `getMatchingFieldTypes`. We support requesting a flattened subfield directly, but do not attempt to resolve wildcard patterns to subfields (this is an explicit choice and noted in docs).

**Review by jtibshirani — APPROVED** [review]
I left one comment about a convenience method `getAllFieldTypes`. After you make a decision with that, it looks ready to merge!

**Pull request merged**